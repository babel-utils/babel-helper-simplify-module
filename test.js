// @flow
'use strict';

const {simplifyModule} = require('./');
const createBabylonOptions = require('babylon-options');
const pluginTester = require('babel-plugin-tester');
const printAST = require('ast-pretty-print');

const plugin = function({types: t}) {
  const VISITED = Symbol('visited');

  let getOutput = path => {
    let printed = printAST(path) + '\n';
    return t.expressionStatement(
      t.templateLiteral([
        t.templateElement({ raw: printed }, true)
      ], [])
    );
  };

  return {
    name: 'test-plugin',
    visitor: {
      Program(path) {
        if (path[VISITED]) return;
        simplifyModule(path);
        path[VISITED] = true;
        path.pushContainer('body', getOutput(path));
      },
    },
  };
};

pluginTester({
  plugin: plugin,
  babelOptions: {
    parserOpts: createBabylonOptions({
      stage: 0,
      plugins: ['flow'],
    }),
  },
  snapshot: true,
  tests: [
    {
      title: 'statement',
      code: 'var a = 1;',
    },
    {
      title: 'statement rewriting',
      code: 'var a = 1, b = 2;',
    },
    {
      title: 'export named',
      code: 'export var a = 1;',
    },
    {
      title: 'export default',
      code: 'export default 1;',
    },
    {
      title: 'export named from',
      code: 'export { a } from "b";',
    },
    {
      title: 'export default from',
      code: 'export default from "b";',
    },
    {
      title: 'export all from',
      code: 'export * from "b";',
    },
    {
      title: 'import default',
      code: 'import a from "b";',
    },
    {
      title: 'import named',
      code: 'import { a } from "b";',
    },
    {
      title: 'import namespace',
      code: 'import * as a from "b";',
    },
    {
      title: 'import type default',
      code: 'import type a from "b";',
    },
    {
      title: 'import type named',
      code: 'import type { a } from "b";',
    },
    {
      title: 'import type inner',
      code: 'import { type a } from "b";',
    },
    {
      title: 'import typeof default',
      code: 'import typeof a from "b";',
    },
    {
      title: 'import typeof named',
      code: 'import typeof { a } from "b";',
    },
    {
      title: 'import typeof inner',
      code: 'import { typeof a } from "b";',
    },
    {
      title: 'import side effects',
      code: 'import "b";',
    },
  ],
});
