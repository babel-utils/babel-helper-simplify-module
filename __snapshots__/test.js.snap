// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1. statement 1`] = `
"
var a = 1;

      ↓ ↓ ↓ ↓ ↓ ↓

var a = 1;\`Path: Program (1:0,1:10)
  body:
    - VariableDeclaration (1:4,1:9)
        declarations:
          - VariableDeclarator (1:4,1:9)
              id: Identifier (1:4,1:5)
                name: 'a'
              init: NumericLiteral (1:8,1:9)
                extra:
                  raw: '1'
                  rawValue: 1
                value: 1
        kind: 'var'
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;

exports[`2. statement rewriting 1`] = `
"
var a = 1, b = 2;

      ↓ ↓ ↓ ↓ ↓ ↓

var a = 1;var b = 2;\`Path: Program (1:0,1:17)
  body:
    - VariableDeclaration (1:4,1:9)
        declarations:
          - VariableDeclarator (1:4,1:9)
              id: Identifier (1:4,1:5)
                name: 'a'
              init: NumericLiteral (1:8,1:9)
                extra:
                  raw: '1'
                  rawValue: 1
                value: 1
        kind: 'var'
    - VariableDeclaration (1:11,1:16)
        declarations:
          - VariableDeclarator (1:11,1:16)
              id: Identifier (1:11,1:12)
                name: 'b'
              init: NumericLiteral (1:15,1:16)
                extra:
                  raw: '2'
                  rawValue: 2
                value: 2
        kind: 'var'
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;

exports[`3. export named 1`] = `
"
export var a = 1;

      ↓ ↓ ↓ ↓ ↓ ↓

var a = 1;export { a };
\`Path: Program (1:0,1:17)
  body:
    - VariableDeclaration (1:11,1:16)
        declarations:
          - VariableDeclarator (1:11,1:16)
              id: Identifier (1:11,1:12)
                name: 'a'
              init: NumericLiteral (1:15,1:16)
                extra:
                  raw: '1'
                  rawValue: 1
                value: 1
        kind: 'var'
    - ExportNamedDeclaration (1:7,1:17)
        declaration: null
        source: null
        specifiers:
          - ExportSpecifier (1:7,1:17)
              exported: Identifier
                name: 'a'
              local: Identifier
                name: 'a'
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;

exports[`4. export default 1`] = `
"
export default 1;

      ↓ ↓ ↓ ↓ ↓ ↓

const _default = 1;export default _default;
\`Path: Program (1:0,1:17)
  body:
    - VariableDeclaration (1:15,1:16)
        declarations:
          - VariableDeclarator (1:15,1:16)
              id: Identifier
                name: '_default'
              init: NumericLiteral (1:15,1:16)
                extra:
                  raw: '1'
                  rawValue: 1
                value: 1
        kind: 'const'
    - ExportDefaultDeclaration (1:0,1:17)
        declaration: Identifier
          name: '_default'
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;

exports[`5. export named from 1`] = `
"
export { a } from \\"b\\";

      ↓ ↓ ↓ ↓ ↓ ↓

export { a } from \\"b\\";
\`Path: Program (1:0,1:22)
  body:
    - ExportNamedDeclaration (1:9,1:10)
        declaration: null
        source: StringLiteral
          value: 'b'
        specifiers:
          - ExportSpecifier (1:9,1:10)
              exported: Identifier
                name: 'a'
              local: Identifier
                name: 'a'
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;

exports[`6. export default from 1`] = `
"
export default from \\"b\\";

      ↓ ↓ ↓ ↓ ↓ ↓

export { default } from \\"b\\";
\`Path: Program (1:0,1:24)
  body:
    - ExportNamedDeclaration (1:7,1:14)
        declaration: null
        source: StringLiteral
          value: 'b'
        specifiers:
          - ExportSpecifier (1:7,1:14)
              exported: Identifier
                name: 'default'
              local: Identifier
                name: 'default'
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;

exports[`7. export all from 1`] = `
"
export * from \\"b\\";

      ↓ ↓ ↓ ↓ ↓ ↓

export * from \\"b\\";
\`Path: Program (1:0,1:18)
  body:
    - ExportAllDeclaration (1:0,1:18)
        source: StringLiteral
          value: 'b'
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;

exports[`8. import default 1`] = `
"
import a from \\"b\\";

      ↓ ↓ ↓ ↓ ↓ ↓

import a from \\"b\\";
\`Path: Program (1:0,1:18)
  body:
    - ImportDeclaration (1:7,1:8)
        importKind: 'value'
        source: StringLiteral
          value: 'b'
        specifiers:
          - ImportDefaultSpecifier (1:7,1:8)
              local: Identifier
                name: 'a'
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;

exports[`9. import named 1`] = `
"
import { a } from \\"b\\";

      ↓ ↓ ↓ ↓ ↓ ↓

import { a } from \\"b\\";
\`Path: Program (1:0,1:22)
  body:
    - ImportDeclaration (1:9,1:10)
        importKind: 'value'
        source: StringLiteral
          value: 'b'
        specifiers:
          - ImportSpecifier (1:9,1:10)
              imported: Identifier
                name: 'a'
              local: Identifier
                name: 'a'
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;

exports[`10. import namespace 1`] = `
"
import * as a from \\"b\\";

      ↓ ↓ ↓ ↓ ↓ ↓

import * as a from \\"b\\";
\`Path: Program (1:0,1:23)
  body:
    - ImportDeclaration (1:7,1:13)
        importKind: 'value'
        source: StringLiteral
          value: 'b'
        specifiers:
          - ImportNamespaceSpecifier (1:7,1:13)
              local: Identifier
                name: 'a'
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;

exports[`11. import type default 1`] = `
"
import type a from \\"b\\";

      ↓ ↓ ↓ ↓ ↓ ↓

import type { a } from \\"b\\";
\`Path: Program (1:0,1:23)
  body:
    - ImportDeclaration (1:12,1:13)
        importKind: 'type'
        source: StringLiteral
          value: 'b'
        specifiers:
          - ImportSpecifier (1:12,1:13)
              imported: Identifier
                name: 'a'
              local: Identifier
                name: 'a'
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;

exports[`12. import type named 1`] = `
"
import type { a } from \\"b\\";

      ↓ ↓ ↓ ↓ ↓ ↓

import type { a } from \\"b\\";
\`Path: Program (1:0,1:27)
  body:
    - ImportDeclaration (1:14,1:15)
        importKind: 'type'
        source: StringLiteral
          value: 'b'
        specifiers:
          - ImportSpecifier (1:14,1:15)
              imported: Identifier
                name: 'a'
              local: Identifier
                name: 'a'
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;

exports[`13. import type inner 1`] = `
"
import { type a } from \\"b\\";

      ↓ ↓ ↓ ↓ ↓ ↓

import type { a } from \\"b\\";
\`Path: Program (1:0,1:27)
  body:
    - ImportDeclaration (1:9,1:15)
        importKind: 'type'
        source: StringLiteral
          value: 'b'
        specifiers:
          - ImportSpecifier (1:9,1:15)
              imported: Identifier
                name: 'a'
              local: Identifier
                name: 'a'
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;

exports[`14. import typeof default 1`] = `
"
import typeof a from \\"b\\";

      ↓ ↓ ↓ ↓ ↓ ↓

import typeof { a } from \\"b\\";
\`Path: Program (1:0,1:25)
  body:
    - ImportDeclaration (1:14,1:15)
        importKind: 'typeof'
        source: StringLiteral
          value: 'b'
        specifiers:
          - ImportSpecifier (1:14,1:15)
              imported: Identifier
                name: 'a'
              local: Identifier
                name: 'a'
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;

exports[`15. import typeof named 1`] = `
"
import typeof { a } from \\"b\\";

      ↓ ↓ ↓ ↓ ↓ ↓

import typeof { a } from \\"b\\";
\`Path: Program (1:0,1:29)
  body:
    - ImportDeclaration (1:16,1:17)
        importKind: 'typeof'
        source: StringLiteral
          value: 'b'
        specifiers:
          - ImportSpecifier (1:16,1:17)
              imported: Identifier
                name: 'a'
              local: Identifier
                name: 'a'
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;

exports[`16. import typeof inner 1`] = `
"
import { typeof a } from \\"b\\";

      ↓ ↓ ↓ ↓ ↓ ↓

import typeof { a } from \\"b\\";
\`Path: Program (1:0,1:29)
  body:
    - ImportDeclaration (1:9,1:17)
        importKind: 'typeof'
        source: StringLiteral
          value: 'b'
        specifiers:
          - ImportSpecifier (1:9,1:17)
              imported: Identifier
                name: 'a'
              local: Identifier
                name: 'a'
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;

exports[`17. import side effects 1`] = `
"
import \\"b\\";

      ↓ ↓ ↓ ↓ ↓ ↓

\`Path: Program (1:0,1:11)
  body: []
  directives: []
  innerComments: []
  leadingComments: []
  sourceType: 'module'
  trailingComments: []
\`;
"
`;
